Create a Type Ahead Feature in JavaScript using Fetch and Regex (#6213)
section-engineering-education/engineering-education: content/articles/create-an-ajax-type-ahead-feature-for-a-website/index.md

---
layout: engineering-education
status: publish
published: true
url: /create-an-ajax-type-ahead-feature-for-a-website/
title: Create a Type Ahead Feature in JavaScript using Fetch and Regex
description: This tutorial will discusss how to build a type-ahead feature for a website using Fetch and Regex.
author: godwin-martins
date: 2022-02-09T00:00:00-10:50
topics: [Languages]
excerpt_separator: <!--more-->
images:

  - url: /engineering-education/create-an-ajax-type-ahead-feature-for-a-website/hero.jpg
    alt: Create a Type Ahead Feature in JavaScript using Fetch and Regex Hero Image
---

> Part of the job of a front-end developer is to create intuitive user interfaces that deliver the best user experience to website users. Part of getting this job done is by creating search suggestions or what is known as the type-ahead feature for websites/webpages.

Reviewer: I feel the start of your article is not very intuitive, kindly rephrase it. You could start by saying this 
` Creating intuitive user interfaces that deliver the best user experience to website users is part of the job of a front-end developer`.   

Kindly avoid being repetitive - the word `part`.

<!--more-->

> This tutorial will explain how to create an **ajax type-ahead feature** for your website using the [cities.json file](). This file contains information about states and cities in America. What this feature does is, if a user types `name`, every state and city that matches the word name pops up.

Reviewer: Add the missing link here.

### Prerequisites
To follow along with this tutorial, the reader should have:
- Basic HTML and CSS knowledge.
- ES6 JavaScript/Asynchronous JavaScript.

> - A Code editor.

Reviewer: ```suggestion
- A code editor.
```

- An [understanding of Regular Expression (RegExp)](https://eloquentjavascript.net/09_regexp.html)

### Table of contents
- [How FETCH works](#How-fetch-works)
- [Getting started](#getting-started)
- [The JavaScript](#the-javascript)
- [Getting the data](#getting-the-data)
- [Displaying data to the UI](#displaying-data-to-the-ui)
- [Formating the output](#formating-the-output)
- [The complete JavaScript code](#the-complete-javascript-code)
- [Conclusion](#conclusion)

### How `fetch` works
`Fetch` like `XMLHttpRequest` provides a method to get resources asynchronously across networks. However, the `Fetch` method gives an easy, cleaner, and more efficient way to handle resources across networks.

This method uses `Promise` to deliver more flexible features to make requests to servers from the web browsers.

_Syntax_

> `Fetch(URL, options)`

Reviewer: Place this method in backticks and identify the language.

### The Javascript
### Getting the data
The data for this project is from the [cities.json](https://gist.githubusercontent.com/Miserlou/c5cd8364bf9b2420bb29/raw/2bf258763cdddd704f8ffd3ea9a3e81d25e2c6f6/cities.json) file, a file that has information of states cities in America.

From the file, we get access to the following information of each state in America:
1. State name
2. City name
3. Population
4. Latitude and longitude
5. Growth_from_2000_to_2013
6. Rank

You can choose the data to display from the list above.

- Get the search word and the `ul` from the UI using the code below:

```js
const searchText = document.querySelector('input[type="search"]');
const card = document.querySelector(".card");
```

- Create the data endpoint and create an empty array as shown below:

```js
const dataEndpoint =
"https://gist.githubusercontent.com/Miserlou/c5cd8364bf9b2420bb29/raw/2bf258763cdddd704f8ffd3ea9a3e81d25e2c6f6/cities.json";

const cities = [];
```

- Now, we need to fetch our data from the endpoint and assign the returned value to the empty array we created.

```js
fetch(dataEndpoint)
  .then((blobdata) => blobdata.json())
  .then((data) => cities.push(...data));
```

> > The blobdata object, in turn, does not contain the actual JSON response body but is instead a representation of the entire HTTP response. To extract the JSON body content from the Response object, we use the `JSON()` method, which returns a second promise that resolves with the result of parsing the response body text as JSON. -[MDN](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)

Reviewer: Is the object supposed to be `blobdata` or `BLOB data`. Could you break down the second sentence here as it is too long. Also, add the MDN link properly.

> ![Blobdata output](./returned-blobdata.png)

Reviewer: ```suggestion
![Blobdata output](/engineering-education/create-an-ajax-type-ahead-feature-for-a-website/returned-blobdata.png)
```
Use the same format for your other images as well. Also, if your images are from third party sources/ are not screenshots, add the image source links.

_blobdata returns entire HTTP response_

Here are a few things to note:
- We defined the empty cities array using a `const` declaration, so we cannot reassign data into it. We can only declare the cities array with a `let` keyword.
- Alternatively, you can push and [spread the data into the cities array](https://www.samanthaming.com/tidbits/92-6-use-cases-of-spread-with-array/).
- Create a function called `getMatches` that takes two arguments, the `wordToMatch`and the `cities` array.

> - In the function, we return a filtered subset of the `cities` array and check to see if what the user typed is returned. To check if what the user typed is inside the array, we will [use Regex](https://www.w3schools.com/js/js_regexp.asp) to check if the searched word match any state or city in the `cities` array.

Reviewer: ```suggestion
 In the function above, we return a filtered subset of the `cities` array and check if what the user typed is returned. To check if what the user typed is inside the array, we will use [Regex](https://www.w3schools.com/js/js_regexp.asp) to check if the searched word matches any state or city in the `cities` array.
```
